third_party/move/move-vm/runtime/src/native_functions.rs

```diff
@@ -209,19 +205,38 @@ impl<'b> NativeContext<'_, 'b> {
         self.interpreter.get_stack_frames(count)
     }
 
-    pub fn gas_balance(&self) -> InternalGas {
-        self.gas_balance
+    pub fn legacy_gas_budget(&self) -> InternalGas {
+        self.gas_meter.legacy_gas_budget_in_native_context()
     }
 
     pub fn use_heap_memory(&mut self, amount: u64) {
-        self.heap_memory_usage = self.heap_memory_usage.saturating_add(amount);
+        self.legacy_heap_memory_usage = self.legacy_heap_memory_usage.saturating_add(amount);
+    }
+
+    pub fn legacy_heap_memory_usage(&self) -> u64 {
+        self.legacy_heap_memory_usage
+    }
+
+    /// Returns the gas meter used for execution. Even if native functions cannot use it to
+    /// charge gas (feature-gating), gas meter can be used to query gas meter's balance.
+    pub fn gas_meter(&mut self) -> &mut dyn NativeGasMeter {
+        self.gas_meter
     }
 
-    pub fn heap_memory_usage(&self) -> u64 {
-        self.heap_memory_usage
+    pub fn charge_gas_for_dependencies(&mut self, module_id: ModuleId) -> VMResult<()> {
```

Comment:
minor and prob. not possibe: where is this implemented for the legacy flow/ can we re-use?

Reply:
We cannot because by this time native context is gone, so we can use it for new flow only :(